src/query/api/v1/handler/placement/replace.go

```diff
@@ -0,0 +1,154 @@
+	M3DBReplaceURL = path.Join(handler.RoutePrefixV1, M3DBServicePlacementPathName, replacePathName)
+)
+
+// ReplaceHandler is the type for placement replaces.
+type ReplaceHandler Handler
+
+// NewReplaceHandler returns a new ReplaceHandler.
+func NewReplaceHandler(opts HandlerOptions) *ReplaceHandler {
+	return &ReplaceHandler{HandlerOptions: opts, nowFn: time.Now}
+}
+
+func (h *ReplaceHandler) ServeHTTP(serviceName string, w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx)
+
+	req, pErr := h.parseRequest(r)
+	if pErr != nil {
+		xhttp.Error(w, pErr.Inner(), pErr.Code())
+		return
+	}
+
+	placement, err := h.Replace(serviceName, r, req)
+	if err != nil {
+		status := http.StatusInternalServerError
+		if _, ok := err.(unsafeAddError); ok {
+			status = http.StatusBadRequest
+		}
+		logger.Error("unable to replace instance", zap.Any("error", err))
+		xhttp.Error(w, err, status)
+		return
+	}
+
+	placementProto, err := placement.Proto()
+	if err != nil {
+		logger.Error("unable to get placement protobuf", zap.Any("error", err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	resp := &admin.PlacementGetResponse{
+		Placement: placementProto,
+		Version:   int32(placement.GetVersion()),
+	}
+
+	xhttp.WriteProtoMsgJSONResponse(w, resp, logger)
+}
+
+func (h *ReplaceHandler) parseRequest(r *http.Request) (*admin.PlacementReplaceRequest, *xhttp.ParseError) {
+	defer r.Body.Close()
+
+	req := &admin.PlacementReplaceRequest{}
+	if err := jsonpb.Unmarshal(r.Body, req); err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	return req, nil
+}
+
+// Replace replaces instances.
+func (h *ReplaceHandler) Replace(
+	serviceName string,
+	httpReq *http.Request,
+	req *admin.PlacementReplaceRequest,
+) (placement.Placement, error) {
+
+	candidates, err := ConvertInstancesProto(req.Candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceOpts := NewServiceOptions(serviceName, httpReq.Header, h.M3AggServiceOptions)
+	service, algo, err := ServiceWithAlgo(h.ClusterClient, serviceOpts, h.nowFn())
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Force {
+		newPlacement, _, err := service.ReplaceInstances(req.LeavingInstanceIDs, candidates)
+		if err != nil {
+			return nil, err
+		}
+
+		return newPlacement, nil
+	}
+
+	curPlacement, version, err := service.Placement()
+	if err != nil {
+		return nil, err
+	}
+
+	// M3Coordinator isn't sharded, can't check if its shards are available.
+	if serviceName != M3CoordinatorServiceName {
+		if err := validateAllAvailable(curPlacement); err != nil {
+			return nil, err
+		}
+	}
+
+	newPlacement, err := algo.ReplaceInstances(curPlacement, req.LeavingInstanceIDs, candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := service.CheckAndSet(newPlacement, version); err != nil {
+		return nil, err
+	}
+
+	return newPlacement.SetVersion(version + 1), nil
```

Comment:
A little annoying that CheckAndSet doesn't set the version, but this is guaranteed to be correct right (version +1)?

Reply:
It currently is but admittedly that's tied to the etcd implementation I think. The API CheckAndSet uses [under the hood](https://github.com/m3db/m3/blob/a2fab37bae1b9144354eba854bcb050a7d6398b3/src/cluster/kv/types.go#L141) propagates the version but we [drop that](https://github.com/m3db/m3/blob/a2fab37bae1b9144354eba854bcb050a7d6398b3/src/cluster/placement/storage/storage.go#L102) when the placement service does CheckAndSet.

We could (and probably should) refactor that but it would likely require updating a bunch of interface implementations. I can open an issue.

Reply:
Hm, you could just do a get placement operation to avoid this possibly?

Reply:
Hey can you return perhaps just:
```go
p, _, err := service.Placement()
return p, err
```

Just so before we have some wrapper that does this for you in a way that can be reused, you at least just read back what you wrote?

Reply:
That read after CAS breaks the atomicity, the current implementation is the only way to get atomicity in placement change with validation, a more proper solution is [this](https://github.com/m3db/m3cluster/issues/243)